c7n/resources/ami.py

```diff
@@ -76,18 +75,24 @@ class Deregister(BaseAction):
                     days: 90
                 actions:
                   - deregister
+                    terminate_snapshot: true
     """
 
-    schema = type_schema('deregister')
+    schema = type_schema('deregister', terminate_snapshot={'type': 'boolean'})
     permissions = ('ec2:DeregisterImage',)
 
     def process(self, images):
-        with self.executor_factory(max_workers=10) as w:
-            list(w.map(self.process_image, images))
+        with self.executor_factory(max_workers=5) as w:
+            w.map(self.process_image, images)
 
     def process_image(self, image):
         client = local_session(self.manager.session_factory).client('ec2')
         client.deregister_image(ImageId=image['ImageId'])
+        if self.data.get('terminate_snapshot', False):
+            snapshot_id = image['BlockDeviceMappings'][0]['Ebs']['SnapshotId']
```

Comment:
this has a few assumptions, first that the ami is for an ebs backed image, not always the case, second that the ami only has one ebs snapshot associated.

Reply:
I'm working on this now! It will delete _all_ snapshots associated with the ami (regardless of if it's ebs or instance-store). and if it's instance-store, it will delete the manifest and the S3 files.